### PR TITLE
Partial ordering

### PR DESCRIPTION
**Pull request by ms32035**
### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by zzzeek** [comment]
heya.  sorry, I'm really out of it.  lets put this in CI and see what happens.   feel free to ping me if I fall asleep again.

**Comment by sqla-tester** [comment]
New Gerrit review created for change 29392b796dd995fabdabe50e8725385dbe1b4883: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/alembic/+/1698

File: tests/test_batch.py [hunk 1]
```diff
@@ -466,6 +466,21 @@ def test_add_col(self):
...
+    def test_add_col_partial_ordering(self):
+        impl = self._simple_fixture()
+        col = Column("g", Integer)
+        # operations.add_column produces a table
+        t = self.op.schema_obj.table("tname", col)  # noqa
+        impl.add_column("tname", col, partial_ordering=("x", "g", "y"))
+        new_table = self._assert_impl(impl, colnames=["id", "x", "g", "y"])
+        eq_(new_table.c.g.name, "g")
+
+    # TODO: doesn't work, don't know why
```
Comment (sqla-tester) [inline tests/test_batch.py]: **mike bayer** (zzzeek) wrote:

so...this passes now and I can take that comment out

View this in Gerrit at https://gerrit.sqlalchemy.org/1698

**Comment by ms32035** [comment]
> **mike bayer** (zzzeek) wrote:
> 
> (1 comment)
> 
> so I think partial_ordering is handy but I bet what you really want for the add_column() use case is a simpler "insert_after" / "insert_before" kind of thing, because partial_ordering doesn't really do that without putting out the whole list of columns, so let me see if that can be done.
> 
> View this in Gerrit at https://gerrit.sqlalchemy.org/1698

well, I guess the difference is that `insert_after/insert_before` requires just one other column assuming the current order is correct, whereas `partial_ordering` needs all the columns, but gives you more power. In my experience (primarily ETL), in most of the cases the former is sufficient, but I've also seen cases when the latter was needed.

**Comment by zzzeek** [comment]
> > **mike bayer** (zzzeek) wrote:
> > (1 comment)
> > so I think partial_ordering is handy but I bet what you really want for the add_column() use case is a simpler "insert_after" / "insert_before" kind of thing, because partial_ordering doesn't really do that without putting out the whole list of columns, so let me see if that can be done.
> > View this in Gerrit at https://gerrit.sqlalchemy.org/1698
> 
> well, I guess the difference is that `insert_after/insert_before` requires just one other column assuming the current order is correct, whereas `partial_ordering` needs all the columns, but gives you more power. In my experience (primarily ETL), in most of the cases the former is sufficient, but I've also seen cases when the latter was needed.

I'm allowing both parameters, but if you want to do total reordering you do it at the batch_alter_table() call.  For the individual add_columns(), my notion that someone could just add a short ordering is too open-ended and doesn't actually work for the most common use case of adding a single column to a specific spot.   It also mirrors your original request to make it behave like MySQL's FIRST / AFTER thing.

**Comment by zzzeek** [comment]
note the latest version, when you give it insert_before or insert_after, there's a whole dump of new logic in there to figure out how to order this new column with the other ones since it has to be specific the way I did this w/ the topological thing.    it's a little overbuilt to support the more open-ended use case.

**Comment by ms32035** [comment]
Had a quick peek at gerrit and it looks amazing. Never seen before any db migration tool to be able to do this in an automated way. Will try out next week. Great stuff. Thanks @zzzeek

**Comment by zzzeek** [comment]
this will close automatically when we're all merged

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

View this in Gerrit at https://gerrit.sqlalchemy.org/1698

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/1698 has been **merged**. Congratulations! :)

**Pull request closed without merging**